### ISSUE TITLE
PayPal IPN Validate fails sometimes (timeout maybe?)

### ISSUE THREAD

**Issue by mircobabini** [strangerstudios/paid-memberships-pro#1787]
**Describe the bug**
I've found a few errors in 1 month of ipn debug logs:

`cURL error 56: OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 104
`
Googling for it I've found may having ALMOST the same issues; they have no "SSL_ERROR_SYSCALL" btw, but they have the connection closed by peer after 5 seconds. wp_remote_post without explicit changes to timeout param uses timeout=5; I've seen others using timeout=60 (ex. WC).

I would suggest to change some `pmpro_ipnValidate` things:
1. pass `$body` param as array, `wp_remote_post` will do the rest
1. `wp_unslash( $body )` which is a deep `stripslashes` replacement
1. use `wp_safe_remote_post`
1. set `'timeout' => 60`

Also, would be great to know why we are using:
- "Host"        => "www.paypal.com",
- "Connection"  => "Close",

**To Reproduce**
Who knows... I've about 100-120 recurring_payment IPN messages every day... and experienced the validate issue twice in a month.

**Comment by mircobabini** [comment]
1st case error: **cURL error 56: OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 104**

```
	ERROR
	Error Info: Array
(
    [http_request_failed] => Array
        (
            [0] => cURL error 56: OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 104
        )

)


	WP_Error Object
(
    [errors] => Array
        (
            [http_request_failed] => Array
                (
                    [0] => cURL error 56: OpenSSL SSL_read: SSL_ERROR_SYSCALL, errno 104
                )

        )

    [error_data] => Array
        (
        )

    [additional_data:protected] => Array
        (
        )

)
```

**Comment by mircobabini** [comment]
2nd case error: **cURL error 28: Operation timed out after 5000 milliseconds with 0 bytes received**

```
	ERROR
	Error Info: Array
(
    [http_request_failed] => Array
        (
            [0] => cURL error 28: Operation timed out after 5000 milliseconds with 0 bytes received
        )

)


	WP_Error Object
(
    [errors] => Array
        (
            [http_request_failed] => Array
                (
                    [0] => cURL error 28: Operation timed out after 5000 milliseconds with 0 bytes received
                )

        )

    [error_data] => Array
        (
        )

    [additional_data:protected] => Array
        (
        )

)
```

**Comment by eighty20results** [comment]
You are correct about why the cURL error is thrown. If the IPN server spends more than 5 seconds waiting, the connection is reset. 

From the standard `error.h` definition:
`#define	ECONNRESET	104	/* Connection reset by peer */`

The challenge with setting a 60 second timeout is that on a very busy system, you run the risk of exhausting the # of sockets available to the site and triggering a much bigger and more complex set of problems. I'd suggest a stepwise climb up in the timeout values, so instead of a massive 20x increase, maybe go for something a bit more conservative? Maybe 10 seconds (if the incoming connections isn't processed in 10 seconds, I would argue there's a general performance issue on the site that should be addressed. Could highlight that to end-users if/when it happens..?

**Comment by mircobabini** [comment]
Thanks for your feedback @eighty20results!

Yeah, why not. We can also make it filterable just in case.

> Could highlight that to end-users if/when it happens..?

I think that the IPN system needs something to notify the admin when this happens; I mean, you can have log enabled or not, but when you face this kind of issues you are missing an order; then probably missing an invoice and so on.

This can be an issue for many european businesses: missing an invoice is something that can lead the business to administrative sanctions in some cases.

A constant like `IPN_NOTIFY_ERRORS` (bool) might be useful as well... or a filter, I don't really like constants btw.

**Comment by mircobabini** [comment]
This happened to me just one more time in 6 months, during a `pmpro_ipnValidate()` call to PayPal:
```
[http_request_failed] => Array
(
    [0] => cURL error 35: OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to www.paypal.com:443
)
```

Closing this for now.

**Comment by dparker1005** [comment]
Thanks for the thorough investigation here. At this point, PayPal Standard and PayPal Express are deprecated gateways in PMPro. PayPal Standard was deprecated in PMPro 3.4, and PayPal Express was deprecated in PMPro 3.7.1.

The IPN retry/history/notification handling proposed here would be new enhancement work for those deprecated integrations, so we’re going to close this as not planned. Sites still relying on these IPN flows should plan to move to a supported payment gateway.
